### PR TITLE
Record that the sigframe write recorded is conservative

### PR DESCRIPTION
**Pull request by KJTsanaktsidis**
When handling a signal in a tracee, the kernel will spill some data to the stack (the "sigframe"), and this data needs to be recorded in the trace for replaying. However, we don't actually know how much data the kernel will write. Instead, we take a conservative estimate which is guaranteed to be at least that size, and record that amount of memory.

However, that range might cross a memory mapping, and during replay, part of that range might not be mapped. Currently that raises an assertion error because the write is smaller than the read.

In this patch, we record whether a memory "write" is actually this kind of conservative capture, and mark it as such in the trace. During replay, we then do not fail that assertion.

Should fix https://github.com/rr-debugger/rr/issues/3779

### CODE REVIEW

**Comment by rocallahan** [comment]
I thought we were going to do your plan, "what if we just made the replay skip the assertion, if the write comes through here"? This looks like an implementation of my plan. I think your plan is better :-).

**Comment by KJTsanaktsidis** [comment]
Oh! Had a closer look at the code when I started implementing today, and I actually thought this way was more elegant. Because the recording code writing the trace is the bit that knows that the memory range recorded is a guess and it may not all be clobbered by the kernel. That was my understanding anyway.

Happy to do it the other way though! I’ll make a different PR.

**Comment by rocallahan** [comment]
No, we can do it this way.

File: src/RecordTask.cc [hunk 1]
```diff
@@ -1724,7 +1724,7 @@ bool RecordTask::record_remote_by_local_map(remote_ptr<void> addr,
   return false;
 }
 
-void RecordTask::record_remote(remote_ptr<void> addr, ssize_t num_bytes) {
+void RecordTask::record_remote(remote_ptr<void> addr, ssize_t num_bytes, bool sizeIsConservative) {
```
Comment (rocallahan) [inline src/RecordTask.cc:1727]: We use snake case, so `size_is_conservative`.

But for public APIs it's better to use a two-value enum than a bool parameter --- more readable.

File: src/TraceStream.h [hunk 2]
```diff
@@ -56,6 +56,7 @@ class TraceStream {
     size_t size;
     pid_t rec_tid;
     std::vector<WriteHole> holes;
+    bool sizeIsConservative;
```
Comment (rocallahan) [inline src/TraceStream.h:59]: snake case

File: src/TraceStream.h [hunk 3]
```diff
@@ -344,6 +347,7 @@ class TraceReader : public TraceStream {
     std::vector<uint8_t> data;
     remote_ptr<void> addr;
     pid_t rec_tid;
+    bool sizeIsConservative;
```
Comment (rocallahan) [inline src/TraceStream.h:350]: Snake case

File: src/TraceStream.h [hunk 4]
```diff
@@ -354,6 +358,7 @@ class TraceReader : public TraceStream {
     remote_ptr<void> addr;
     pid_t rec_tid;
     std::vector<WriteHole> holes;
+    bool sizeIsConservative;
```
Comment (rocallahan) [inline src/TraceStream.h:361]: snake case

File: src/TraceStream.cc [hunk 5]
```diff
@@ -1237,8 +1237,9 @@ KernelMapping TraceReader::read_mapped_region(MappedData* data, bool* found,
 
 void TraceWriter::write_raw_header(pid_t rec_tid, size_t total_len,
                                    remote_ptr<void> addr,
-                                   const std::vector<WriteHole>& holes = std::vector<WriteHole>()) {
-  raw_recs.push_back({ addr, total_len, rec_tid, holes });
+                                   const std::vector<WriteHole>& holes,
+                                   bool sizeIsConservative) {
+  raw_recs.push_back({ addr, total_len, rec_tid, holes, sizeIsConservative});
```
Comment (rocallahan) [inline src/TraceStream.cc:1242]: space before }

File: src/TraceStream.cc [hunk 6]
```diff
@@ -573,7 +573,7 @@ TraceFrame TraceReader::read_frame(FrameTime skip_before) {
       const auto& hole = holes[j];
       h[j] = { hole.getOffset(), hole.getSize() };
     }
-    raw_recs[i] = { w.getAddr(), (size_t)w.getSize(), i32_to_tid(w.getTid()), h };
+    raw_recs[i] = { w.getAddr(), (size_t)w.getSize(), i32_to_tid(w.getTid()), h, w.getSizeIsConservative()};
```
Comment (rocallahan) [inline src/TraceStream.cc:576]: Somewhere there should be a call to `setSizeIsConservative()` but I think it's missing.

**Review by rocallahan — CHANGES_REQUESTED** [review]

File: src/TraceStream.cc (hunk 6 above)
Comment (KJTsanaktsidis) [reply src/TraceStream.cc]: Uh, whoops, yeah, this is very obviously not going to do anything useful without it.

File: src/RecordTask.cc (hunk 1 above)
Comment (KJTsanaktsidis) [reply src/RecordTask.cc]: I wasn't 100% sure what a good name for this would be. `MemWriteSizeValidaiton` seems a bit wordy, but there it is.

**Comment by rocallahan** [comment]
There's no easy way to test this so I think we should just ignore that for now.

**Review by rocallahan — APPROVED** [review]
